### PR TITLE
feat: adding universe domain support for downscroped credentials

### PR DESCRIPTION
**Pull request by BigTailWolf**

### CODE REVIEW

File: google/auth/credentials.py [hunk 1]
```diff
@@ -24,6 +24,8 @@
 from google.auth import metrics
 from google.auth._refresh_worker import RefreshThreadManager
 
+DEFAULT_UNIVERSE_DOMAIN = "googleapis.com"
```
Comment (lsirac) [inline google/auth/credentials.py:27]: I think there are other instances of this, e.g. in external account. Can you consolidate? 

Also, "googleapis.com" is used below. Should be replaced there & elsewhere

File: tests/test_downscoped.py [hunk 2]
```diff
@@ -496,6 +504,20 @@ def test_default_state(self):
         assert not credentials.expired
         # No quota project ID set.
         assert not credentials.quota_project_id
+        assert credentials.universe_domain == DEFAULT_UNIVERSE_DOMAIN
+
+    def test_create_with_customized_universe_domain(self):
```
Comment (lsirac) [inline tests/test_downscoped.py:509]: Can you add more tests? e.g. that request is sent to STS with the custom universe domain

File: google/auth/downscoped.py [hunk 3]
```diff
@@ -467,7 +471,10 @@ def __init__(
         self._source_credentials = source_credentials
         self._credential_access_boundary = credential_access_boundary
         self._quota_project_id = quota_project_id
-        self._sts_client = sts.Client(_STS_TOKEN_URL)
+        self._universe_domain = universe_domain or credentials.DEFAULT_UNIVERSE_DOMAIN
```
Comment (lsirac) [inline google/auth/downscoped.py:474]: Can you check that the universe domain matches the one in source creds?

**Review by lsirac — CHANGES_REQUESTED** [review]

File: google/auth/downscoped.py [hunk 4]
```diff
@@ -437,7 +437,11 @@ class Credentials(credentials.CredentialsWithQuotaProject):
     """
 
     def __init__(
-        self, source_credentials, credential_access_boundary, quota_project_id=None
+        self,
+        source_credentials,
+        credential_access_boundary,
+        quota_project_id=None,
+        universe_domain=credentials.DEFAULT_UNIVERSE_DOMAIN,
```
Comment (arithmetic1728) [inline google/auth/downscoped.py:444]: add a doc string for `universe_domain` in the `Args` part below?

**Review by arithmetic1728 — APPROVED** [review]

File: google/auth/credentials.py (hunk 1 above)
Comment (BigTailWolf) [reply google/auth/credentials.py:27]: Great. In the beginning I wanted to do those refactoring together and I realized the PR is big.
Then i planned to open a PR only to refactor the `DEFAULT_UNIVERSE_DOMAIN` at once.

File: google/auth/downscoped.py (hunk 3 above)
Comment (BigTailWolf) [reply google/auth/downscoped.py:474]: It is covered in the test L#507
```
        assert credentials.universe_domain == DEFAULT_UNIVERSE_DOMAIN
```

Comment (lsirac) [reply google/auth/downscoped.py:474]: What I mean here is if you pass source creds with universe domain A, but pass universe domain B here. We should error out in this case

Comment (BigTailWolf) [reply google/auth/downscoped.py:474]: I think the test supposed to be:
1. Implicit default. Pass nothing for universe domain and validate the universe domain as the default.
2. Explicitly pass None. And valid the universe domain is equals to default.
3. Explicitly pass a universe domain A, and validate the universe domain equals to A

Now working on the test above. Any missing validation?

File: tests/test_downscoped.py [hunk 5]
```diff
@@ -506,6 +543,49 @@ def test_with_quota_project(self):
 
         assert quota_project_creds.quota_project_id == "project-foo"
 
+    @mock.patch("google.auth._helpers.utcnow", return_value=datetime.datetime.min)
+    def test_refresh_on_tpc_universe(self, unused_utcnow):
```
Comment (lsirac) [inline tests/test_downscoped.py:547]:
```suggestion
    def test_refresh_on_custom_universe(self, unused_utcnow):
```

**Review by lsirac — APPROVED** [review]

**Pull request merged**